maven-plugin/src/test/java/org/sonatype/nexus/maven/staging/StagingUploadMojoTest.java

```diff
@@ -0,0 +1,268 @@
+  }
+
+  @Test
+  public void saveTagToPropertiesFile() throws Exception {
+    underTest.execute();
+
+    File directory = underTest.getWorkDirectoryRoot();
+
+    String props = readFileToString(new File(directory, "staging/staging.properties"));
+
+    assertThat(props, containsString("staging.tag=" + TAG));
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void throwIllegalArgumentExceptionWhenIncorrectServerId() throws Exception {
+    when(settings.getServer(anyString())).thenReturn(null);
+
+    underTest.execute();
+  }
+
+  @Test
+  public void getServerConfiguration() {
+    ServerConfig config = underTest.getServerConfiguration(session);
+
+    assertThat(config, is(notNullValue()));
+  }
+
+  @Test
+  public void readIndexFile() {
+    underTest.setStagingIndexFilename("example.index");
+    underTest.setAltStagingDirectory(new File(getBasedir(), "src/test/resources/"));
+    List<ArtifactInfo> artifacts = underTest.readStoredArtifactsFromIndex();
+    assertThat(artifacts.size(), is(2));
+  }
+
+  @Test
+  public void uploadArtifacts() throws MojoExecutionException, MojoFailureException, RepositoryManagerException {
+    underTest.setStagingIndexFilename("example.index");
+    underTest.setAltStagingDirectory(new File(getBasedir(), "src/test/resources/"));
+    underTest.execute();
+
+    ArgumentCaptor<Component> componentArgumentCaptor = ArgumentCaptor.forClass(Component.class);
+    verify(client).upload(any(), componentArgumentCaptor.capture(), eq(TAG));
+    assertThat(componentArgumentCaptor.getValue().getAssets().size(), is(2));
+  }
+
+  private StagingUploadMojo lookupMojo() throws Exception {
+    File testPom = getPom();
+    StagingUploadMojo mojo = (StagingUploadMojo) lookupMojo("upload", testPom);
+
+    mojo.setMavenSession(session);
+    mojo.setArtifact(artifact);
+    mojo.setTag(TAG);
+    mojo.setTagGenerator(tagGenerator);
+    mojo.setClientFactory(clientFactory);
+    return mojo;
+  }
+
+  private File getPom() {
+    return new File(getBasedir(), "src/test/resources/example-pom.xml");
+  }
+
+  private void setupMockBehaviour() throws Exception {
+    Server server = new Server();
+    server.setUsername(USERNAME);
+    server.setPassword(PASSWORD);
+
+    userProperties = new Properties();
+
+    when(session.getExecutionRootDirectory()).thenReturn(tempDirectory.toString());
+
+    when(session.getSettings()).thenReturn(settings);
+
+    when(session.getUserProperties()).thenReturn(userProperties);
+
+    when(settings.getServer(anyString())).thenReturn(server);
+
+    mockArtifact(artifact);
+    mockArtifact(attachedArtifact);
+
+    when(tagGenerator.generate(ARTIFACT_ID, VERSION)).thenReturn(GENERATED_TAG);
+
+    when(clientFactory.build(any())).thenReturn(client);
+
+    when(client.getTag(TAG)).thenReturn(Optional.of(new Tag(TAG)));
+  }
+
+  private void mockArtifact(final Artifact artifact) {
+    when(artifact.getGroupId()).thenReturn(GROUP_ID);
```

Comment:
instead of mocking this it looks like there is a DefaultArtifact implementation you could use.

Reply:
DefaultArtifact does have some quirks (there's an ArtifactHandler within it) but a good suggestion and I'll see if this can be replaced as a drop-in.

Reply:
Were you unable to do this?

Reply:
Done in [ead71ff](https://github.com/sonatype/nxrm3-maven-plugin/pull/36/commits/ead71ffee2f92e843ac341fe6d8de9bf3cbce424)